torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+
+  // keep track of one sumToSize hitting each output
+  auto& outputMapAndSizes = spec.outputMapAndSizes();
+  JIT_ASSERT(outputMapAndSizes.empty());
+  std::vector<int64_t> outputGradSumToSizes(graph->outputs().size(), -1);
+
+  static OperatorSet commutes_with_SumToSize{{
+      "aten::mul(Tensor self, Tensor other) -> Tensor",
+      "aten::div(Tensor self, Tensor other) -> Tensor",
+      // for div we might check whether we're the first argument
+      "aten::mul(Tensor self, Scalar other) -> Tensor",
+      "aten::div(Tensor self, Scalar other) -> Tensor",
+      "aten::neg(Tensor self) -> Tensor",
+      "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
+      // add this used to be prim::AutogradAdd
+  }};
+  // Scan the graph. We will delete nodes. We want later (in the graph)
+  // _grad_sum_to_size nodes to have priority over earlier ones. Thus
+  // we scan backwards.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend(); it++) {
+    auto* node = *it;
+    if (node->kind() == aten::_grad_sum_to_size) {
+      use_list uses_to_process(node->output()->uses());
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
+        if (user->matches(
+                "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
+          // sometimes, a mask or similar is cast to the same
+          // type as the gradient. But we only want to allow that as offset==1
+          // (i.e. the type)
+          JIT_ASSERT(offset == 1);
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (commutes_with_SumToSize.find(user)) {
+          // these are expressions that might occur during autotdiff operating
+          // on the gradient (matmul would likely be, too but we don't fuse it)
+          // note that for mul, we know (from the chain rule) that only one
+          // factor will be stemming from a calculation involving gradients so
+          // we know that we can move SumToSize across it
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // this case should not happen due to how we organize the search
+          // and remove nodes
+          JIT_ASSERT(false);
+        } else if (user->kind() == prim::Return) {
+          // only if we don't already have a _grad_sum_to_size for this,
+          // because we want later (in the graph) nodes to be preferrable
+          // to earlier ones and we iterate backwards
+          // The implicit assumption is that if we have several _grad_sumtosizes
+          // "in parallel" (from addition) they are the same.
```

Comment:
I don't get the part about "from addition". Can you please elaborate?

Reply:
If you have the backward of `y = torch.sin(phi)**2 + torch.cos(phi)**2`, you'll have `dy/dphi = something_from_sin + something_from_cos`. Now if you do different broadcasting or do something where the fusion group has several outputs, you have several inputs. However, given that the final sumtosize would be to the shape of phi, they should all be the same. (Of course, in practice, the gradient might not always as easy to compute as in the above example.)